velero-csi-plugin/volumesnapshot_restorer.go

```diff
@@ -31,29 +32,50 @@ type VSRestorer struct {
 	log logrus.FieldLogger
 }
 
+// AppliesTo returns information about which resources the VSRestorer action should be invoked for.
+// VSRestorer RestoreItemAction plugin's Execute function will only be invoked on items that match the returned
+// selector. A zero-valued ResourceSelector matches all resources.
 func (p *VSRestorer) AppliesTo() (velero.ResourceSelector, error) {
 	return velero.ResourceSelector{
 		IncludedResources: []string{"volumesnapshots.snapshot.storage.k8s.io"},
 	}, nil
 }
 
+// Execute allows the RestorePlugin to perform arbitrary logic with the item being restored,
+// in this case, logic to correctly restore a CSI VolumeSnapshot custom resource that represents a snapshot of a CSI backed volume.
 func (p *VSRestorer) Execute(input *velero.RestoreItemActionExecuteInput) (*velero.RestoreItemActionExecuteOutput, error) {
-	p.log.Info("Starting VSRestorer")
+	p.log.Info("Starting VSRestorerAction")
 	var vs snapshotv1beta1api.VolumeSnapshot
 
 	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(input.Item.UnstructuredContent(), &vs); err != nil {
-		return &velero.RestoreItemActionExecuteOutput{}, err
+		return &velero.RestoreItemActionExecuteOutput{}, errors.Wrapf(err, "failed to convert input.Item from unstructured")
+	}
+
+	var vsFromBackup snapshotv1beta1api.VolumeSnapshot
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured((input.ItemFromBackup.UnstructuredContent()), &vsFromBackup); err != nil {
+		return &velero.RestoreItemActionExecuteOutput{}, errors.Wrapf(err, "failed to convert input.ItemFromBackup from unstructured")
 	}
 
-	// This section of code used to set the VolumeSnapshot.Spec.Source field to nil
-	// As of the beta, this is no longer relevant, and the code has been removed.
+	if vsFromBackup.Status == nil || vsFromBackup.Status.BoundVolumeSnapshotContentName == nil {
+		return &velero.RestoreItemActionExecuteOutput{}, errors.Errorf("unable to lookup BoundVolumeSnapshotContentName from status")
+	}
+
+	vscName := *vsFromBackup.Status.BoundVolumeSnapshotContentName
```

Comment:
this line seems unnecessary since we want a string pointer anyway on L69

Reply:
Right. I make this mistake of using C like `strdup`, bc the pointer may be freed under me, to get my own pointer instead of reusing it! Here we share the pointers 😄